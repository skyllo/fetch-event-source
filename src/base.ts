/* eslint-disable no-underscore-dangle */
/* eslint-disable no-await-in-loop */

// ready state
const CONNECTING = 0;
const OPEN = 1;
const CLOSED = 2;

export interface FetchEventSourceConfiguration {
  /** custom fetch implementation */
  fetch?: typeof fetch;
  /** custom headers to add to SSE request */
  headers?: Headers;
  /** automatically starts the SSE request */
  autoStart?: boolean;
  /** send cookies to cross-origin URLs */
  withCredentials?: boolean;
  /** initial reconnection delay */
  reconnectionDelay?: number;
}

export default abstract class BaseFetchEventSource extends EventTarget {
  public url: string;

  public withCredentials: boolean;

  public readyState: number = CONNECTING;

  public onopen: () => void | undefined;

  public onmessage: (e?: MessageEvent) => void | undefined;

  public onerror: (error?: ErrorEvent) => void | undefined;

  protected abstract _fetch: typeof fetch | any;

  protected abstract _decode: (bytes) => string;

  protected abstract _getOrigin: () => string;

  protected _transformStream = (stream) => stream;

  #controller: AbortController = new AbortController();

  #config: FetchEventSourceConfiguration = {
    headers: new Headers(),
    autoStart: true,
    reconnectionDelay: 2000,
    withCredentials: false,
  };

  #lastId = '';

  constructor(url: string, config?: FetchEventSourceConfiguration) {
    super();
    // check url is set
    if (!url) {
      throw new Error('Cannot open an FetchgEventSource to an empty URL.');
    }
    this.url = url;
    this.#config = { ...this.#config, ...config };
    this.withCredentials = this.#config.withCredentials;
    // if autoStart set then start connection
    if (this.#config.autoStart) {
      // start on nextTick so everything is initialized
      setTimeout(this.start.bind(this), 0);
    }
  }

  public async start() {
    try {
      await this._connect();
    } catch (e) {
      // start new connection if user did not abort
      if (!this.#controller.signal.aborted) {
        this._error(e);
        await new Promise((resolve) => setTimeout(resolve, this.#config.reconnectionDelay));
        this.start();
      }
    }
  }

  private async _connect() {
    this.#controller = new AbortController();
    // setup fetch and headers
    const { fetch = this._fetch, headers } = this.#config;
    headers.set('Accept', 'text/event-stream');
    if (this.#lastId) {
      headers.set('Last-Event-ID', this.#lastId);
    }
    // connecting
    this.readyState = CONNECTING;
    // make fetch request to SSE
    const response = await fetch(this.url, {
      signal: this.#controller.signal,
      headers,
      credentials: this.#config.withCredentials ? 'include' : 'same-origin',
      cache: 'no-store',
    });
    // fail if not 200 response
    if (!response || response.status !== 200) {
      this.close();
      this._error(new Error('Bad status'));
      return;
    }
    // check is correct content-type from response
    const type = response.headers.get('content-type');
    if (!type || type.split(';')[0] !== 'text/event-stream') {
      this.close();
      this._error(new Error('Bad content-type'));
      return;
    }
    // connected
    this.readyState = OPEN;
    this.dispatchEvent(new Event('open'));
    if (this.onopen) this.onopen();
    // transform stream and process reader
    const reader = this._transformStream(response.body).getReader();
    await this._process(reader);
  }

  private async _process(reader) {
    let type: string = '';
    let data: string[] = [];
    let buffer: string = '';
    for (;;) {
      // stop if has been aborted
      if (this.#controller.signal.aborted) {
        reader.cancel();
        return;
      }
      // check if buffer contains CRLF, LF or CR
      const regex = /\r\n|\r|\n/.exec(buffer);
      if (regex) {
        // get next line which can be a new line
        const line = buffer.slice(0, regex.index) || regex[0];
        buffer = buffer.slice(regex.index + regex[0].length);
        const [key, value = ''] = line.split(/:(.+)?/, 2).map((v2) => v2?.trimLeft());
        // send event if line is a new line
        if (line.match(/^(\r\n|\r|\n)$/) && (type || data)) {
          // create event
          const event = new MessageEvent(type, {
            origin: this._getOrigin(),
            data: data.join('\n'),
            lastEventId: this.#lastId,
          });
          // dispatch event
          this.dispatchEvent(event);
          if (type === 'message' && this.onmessage) this.onmessage(event);
          // reset event data
          type = '';
          data = [];
        }
        // process line
        switch (key) {
          case 'event': {
            type = value;
            break;
          }
          case 'data': {
            type = type || 'message';
            data.push(value);
            break;
          }
          case 'id': {
            this.#lastId = value;
            break;
          }
          case 'retry': {
            const num = Number(value);
            if (num) this.#config.reconnectionDelay = num;
            break;
          }
          default: break;
        }
      } else {
        // read next data
        const { value: byteValue, done } = await reader.read();
        if (done) throw Error('Connection terminated');
        // decode value for both browsers and node
        const value = (typeof byteValue === 'string') ? byteValue : this._decode(byteValue);
        buffer += value;
      }
    }
  }

  close() {
    this.readyState = CLOSED;
    this.#controller.abort();
  }

  _error(error) {
    const event = new ErrorEvent('error', { error, message: error.message });
    this.dispatchEvent(event);
    if (this.onerror) this.onerror(event);
  }
}
