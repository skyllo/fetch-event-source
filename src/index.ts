import isAbsoluteURL from 'is-absolute-url';
import BaseFES from './base';

const decoder = new TextDecoder();

export default class FetchEventSource extends BaseFES {
  _fetch = window.fetch.bind(window);

  _decode = (bytes) => decoder.decode(bytes);

  _getOrigin = () => {
    if (isAbsoluteURL(this.url)) {
      return new URL(this.url).origin;
    }
    return window.location.origin;
  }
}