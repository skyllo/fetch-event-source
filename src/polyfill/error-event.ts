import Event from './event';

export default class ErrorEvent extends Event {
  message: any;

  error: any;

  constructor(type: any, { message, error }) {
    super(type);
    this.message = message;
    this.error = error;
  }
}
