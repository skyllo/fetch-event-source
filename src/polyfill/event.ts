export default class Event {
  type: string;

  constructor(type: any) {
    this.type = type;
  }
}
