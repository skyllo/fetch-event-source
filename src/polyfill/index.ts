/* eslint-disable max-classes-per-file */
import EventTarget from 'event-target-shim';
import AbortController from 'abort-controller';
import { Headers } from 'node-fetch';
import Event from './event';
import ErrorEvent from './error-event';
import MessageEvent from './message-event';

global.AbortController = AbortController;
global.EventTarget = EventTarget;
// @ts-ignore
global.Headers = Headers;
// @ts-ignore
global.Event = Event;
// @ts-ignore
global.MessageEvent = MessageEvent;
// @ts-ignore
global.ErrorEvent = ErrorEvent;
