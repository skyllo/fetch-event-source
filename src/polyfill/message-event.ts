import Event from './event';

export default class MessageEvent extends Event {
  lastEventId: string;

  origin: string;

  data: any;

  constructor(type: any, { origin, data, lastEventId }) {
    super(type);
    this.lastEventId = lastEventId;
    this.origin = origin;
    this.data = data;
  }
}
