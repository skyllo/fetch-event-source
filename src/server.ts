import fetch from 'node-fetch';
import { ReadableStream } from 'web-streams-polyfill/ponyfill';
import { TextDecoder } from 'util';
import BaseFES from './base';

function readableStreamFromAsyncIterable(iterable) {
  const it = iterable[Symbol.asyncIterator]();
  return new ReadableStream({
    async pull(controller) {
      const { done, value } = await it.next();
      if (done) {
        controller.close();
      } else {
        controller.enqueue(value);
      }
    },
    async cancel(reason) {
      await it.throw(reason);
    },
  }, { highWaterMark: 0 });
}

const decoder = new TextDecoder();

export default class FetchEventSource extends BaseFES {
  _fetch = fetch;

  _decode = (bytes) => decoder.decode(bytes);

  _getOrigin = () => new URL(this.url).origin;

  _transformStream = readableStreamFromAsyncIterable;
}
