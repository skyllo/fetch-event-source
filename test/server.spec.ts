/* eslint-disable import/first */
jest.mock('node-fetch');

import { PassThrough } from 'stream';
import fetch from 'node-fetch';
import '../src/polyfill';
import FetchEventSource from '../src/server';

const { Headers, Response } = jest.requireActual('node-fetch');
global.Headers = Headers;

// create a SSE
function createEvent(eventName, data, id = Date.now()) {
  return `id: ${id}\nevent: ${eventName}\ndata: ${JSON.stringify(data)}\n\n`;
}

const mockedFetch = fetch as jest.MockedFunction<typeof fetch>;

// mock response to return SSE
function setupMockedResponse(eventString: string) {
  // mock events
  const mockedStream = new PassThrough();
  mockedStream.push(eventString);
  mockedStream.end();
  // mock response
  const response = new Response(mockedStream, { headers: { 'content-type': 'text/event-stream' } });
  mockedFetch.mockReturnValue(Promise.resolve(response));
}

beforeEach(() => {
  mockedFetch.mockClear();
});

it('should correctly trigger events when data is split', async (done) => {
  // mock events
  const NUM_EVENTS = 5;
  const mockedStream = new PassThrough();
  const data = Array(NUM_EVENTS)
    .fill('')
    .map((_, index) => createEvent('eventName', `secrets${index}`, index))
    .join('')
    .match(/(.|[\r\n]){1,4}/g);
  data.forEach((split) => mockedStream.push(split));
  mockedStream.end();
  // mock response
  const response = new Response(mockedStream, { headers: { 'content-type': 'text/event-stream' } });
  mockedFetch.mockReturnValue(Promise.resolve(response));
  // start fetcheventsource
  const eventSource = new FetchEventSource('https://lol.com');
  const events = [];
  eventSource.addEventListener('eventName', (e) => {
    events.push(e);
    if (events.length === NUM_EVENTS) {
      eventSource.close();
      expect(events).toMatchSnapshot();
      done();
    }
  });
});

it('should handle multiple data fields in one event (SPEC Example 1)', async (done) => {
  setupMockedResponse(`
data: YHOO
data: +2
data: 10

`);
  // start fetcheventsource
  const eventSource = new FetchEventSource('https://lol.com');
  eventSource.addEventListener('message', (e) => {
    eventSource.close();
    expect(e).toMatchSnapshot();
    done();
  });
});

it('should handle comment fields, spaces and empty ID values (SPEC Example 2)', async (done) => {
  setupMockedResponse(`
: test stream

data: first event
id: 1

data:second event
id:

data:  third event

`);
  // start fetcheventsource
  const eventSource = new FetchEventSource('https://lol.com');
  const events = [];
  eventSource.addEventListener('message', (e) => {
    events.push(e);
    if (events.length === 3) {
      eventSource.close();
      expect(events).toMatchSnapshot();
      done();
    }
  });
});

it('should only send events after new lines and double empty data should fire event with a new line (SPEC Example 3)', async (done) => {
  setupMockedResponse(`
data

data
data

data:
`);
  // start fetcheventsource
  const eventSource = new FetchEventSource('https://lol.com');
  const events = [];
  eventSource.addEventListener('message', (e) => {
    events.push(e);
    if (events.length === 2) {
      eventSource.close();
      expect(events).toMatchSnapshot();
      done();
    }
  });
});

it('should send ignore white space after colon to produce same event (SPEC Example 4)', async (done) => {
  setupMockedResponse('data:test\n\ndata: test\n\n');
  // start fetcheventsource
  const eventSource = new FetchEventSource('https://lol.com');
  const events = [];
  eventSource.addEventListener('message', (e) => {
    events.push(e);
    if (events.length === 2) {
      eventSource.close();
      expect(events).toMatchSnapshot();
      done();
    }
  });
});
