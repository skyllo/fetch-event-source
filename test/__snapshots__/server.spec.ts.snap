// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly trigger events when data is split 1`] = `
Array [
  Event {
    "data": "\\"secrets0\\"",
    "isTrusted": false,
    "lastEventId": "0",
    "origin": "https://lol.com",
  },
  Event {
    "data": "\\"secrets1\\"",
    "isTrusted": false,
    "lastEventId": "1",
    "origin": "https://lol.com",
  },
  Event {
    "data": "\\"secrets2\\"",
    "isTrusted": false,
    "lastEventId": "2",
    "origin": "https://lol.com",
  },
  Event {
    "data": "\\"secrets3\\"",
    "isTrusted": false,
    "lastEventId": "3",
    "origin": "https://lol.com",
  },
  Event {
    "data": "\\"secrets4\\"",
    "isTrusted": false,
    "lastEventId": "4",
    "origin": "https://lol.com",
  },
]
`;

exports[`should handle comment fields, spaces and empty ID values (SPEC Example 2) 1`] = `
Array [
  Event {
    "data": "first event",
    "isTrusted": false,
    "lastEventId": "1",
    "origin": "https://lol.com",
  },
  Event {
    "data": "second event",
    "isTrusted": false,
    "lastEventId": "",
    "origin": "https://lol.com",
  },
  Event {
    "data": "third event",
    "isTrusted": false,
    "lastEventId": "",
    "origin": "https://lol.com",
  },
]
`;

exports[`should handle multiple data fields in one event (SPEC Example 1) 1`] = `
Event {
  "data": "YHOO
+2
10",
  "isTrusted": false,
  "lastEventId": "",
  "origin": "https://lol.com",
}
`;

exports[`should only send events after new lines and double empty data should fire event with a new line (SPEC Example 3) 1`] = `
Array [
  Event {
    "data": "",
    "isTrusted": false,
    "lastEventId": "",
    "origin": "https://lol.com",
  },
  Event {
    "data": "
",
    "isTrusted": false,
    "lastEventId": "",
    "origin": "https://lol.com",
  },
]
`;

exports[`should send ignore white space after colon to produce same event (SPEC Example 4) 1`] = `
Array [
  Event {
    "data": "test",
    "isTrusted": false,
    "lastEventId": "",
    "origin": "https://lol.com",
  },
  Event {
    "data": "test",
    "isTrusted": false,
    "lastEventId": "",
    "origin": "https://lol.com",
  },
]
`;
